Тесты сетевого слоя

aitap opened the pull request:
Используются только входящие в комплект Perl модули, но на CentOS7 придётся сделать `yum install perl-Test-Simple`.

Использование: из любой разумной директории запустить `perl .../путь/к/test/network/test.pl`. Скрипт предполагает, что в корне репозитория есть директория (обычно `build` или `cmake-build-debug`), внутри которой лежит исполняемый файл `src/afina`.

Можно добавить параметры командной строки:
 * `-a` / `--afina=путь/к/src/afina` - позволяет передать путь к программе вручную, если не удалось найти его автоматически
 * `-b` / `--backend=...` передаёт afina параметр `--network=...`, по умолчанию `blocking`
 * `-s` / `--silent` отключает печать того, что сама afina выводит на экран.

Скрипт печатает очень много информации, в частности, все отправленные и принятые по сети строки. Для проваленных тестов также выводится на экран ожидаемый (но не полученный) ответ. Тесты команды `replace` проваливаются, но помечаются `TODO`, поэтому не считаются за ошибки.

Совет: исправление ошибок лучше всего начинать с самого первого проваленного теста, потому что последующие ошибки могли быть вызваны им самим (например, упавшая afina приводит к феерической серии ошибок и роняет сам скрипт на половине).

aitap committed:
network tests: try harder to find afina

aitap committed:
network tests: test for errors, too

aitap committed:
network tests: test for actually observable errors

aitap committed:
network tests: test for actually observable errors

aitap committed:
network tests: final touches

fix number of tests, fix glob truthness warning, make script executable

aitap committed:
last minute fix: actually use command line if cannot find src/afina

xphoenix commented:
Круто! А можно переместить тест в itest/network_test.pl (itest - integration tests), тогда я потом поменяю test->utest (uint tests)

Кроме того, есть такая штука как [TAP тесты](https://www.effectiveperlprogramming.com/2011/05/understand-the-test-anything-protocol/), если переделаешь на них будет еще круче. Довольно популярная нынче штука. Главное, что вывод проще интегрировать с различными тулзами и IDE

aitap committed:
move the network test to integration test directory

aitap committed:
network test: use note() instead of diag()

so that `prove` TAP driver would be able to silence the output
(use -v to show it)

aitap committed:
make network test easier to use with TAP driver

 - use environment variables NETWORK_BACKEND and AFINA_PATH to
   supply parameters when command line isn't available
 - don't `note` network interaction when -s is enabled (to mimic prove)

aitap commented:
> А можно переместить тест в itest/network_test.pl

Готово.

> Кроме того, есть такая штука как TAP тесты

Ага, я как раз пользуюсь модулем, который печатает вывод в формате TAP. Сделал вывод дополнительной информации таким образом, чтобы TAP-драйверы могли его спрятать.

Дополнение к инструкции:

Можно также запустить `prove .../путь/к/itest/network_test.pl`, он соберёт краткую статистику и отдельно выведет красным цветом проваленные тесты. Аргументы командной строки тогда нужно будет отделить при помощи `::`, например, `prove itest/network_test.pl :: -a ../build/src/afina`.

Если TAP-драйвер не позволяет передавать аргументы командной строки, можно назначить переменные окружения `AFINA_PATH` и `NETWORK_BACKEND`.
